### ISSUE TITLE
[Question] Using transfer learning to improve ASR on small corpora

### ISSUE THREAD

**Issue by lnalborczyk** [espnet/espnet#2956]
I would like to train an ASR (encoder-decoder) network on a very small French corpus containing around 100 sentences. To help with this task, I would like to first train a similar ASR network on a larger corpus (in French or in another language, cf. for instance the MSL corpus, https://arxiv.org/abs/2012.03411), to freeze the decoder's weights, and to use them to help with the decoding of the smaller corpus.

I have seen some threads about transfer learning with espnet2 (e.g., #2667), but would that be doable with espnet1? If yes, is there any example of a similar procedure?

**Comment by b-flo** [comment]
Hi,

For transfer learning and freezing in ESPnet1, please refer to: https://espnet.github.io/espnet/tutorial.html#how-to-use-finetuning

**Comment by lnalborczyk** [comment]
Perfect, thank you!

**Comment by lnalborczyk** [comment]
I am not sure I understand how to apply freezing and/or transfer learning in practice from the documentation alone. Is it as simple as executing `./run.sh --dec-init-mods=model.loss.best`?

Is there any example in the recipes?

**Comment by b-flo** [comment]
In general, options are passed to `asr_train.py` through training configs. Basically to transfer all modules weights from a pre-trained decoder you should put in your yaml file:

```
dec-init: [model/snapshot_path]
dec-init-mods: [list of modules to transfer, separated by a comma]
```

**Comment by lnalborczyk** [comment]
Thank you very much for your detailed answer! It helps a lot.

Similarly, is it possible to use a pre-trained LM during decoding? Is there any implementation example?

**Comment by b-flo** [comment]
Sorry, do you mean using a LM during decoding or using a pre-trained LM to finetune a model part (e.g.: transducer decoder/prediction net)?

For the first question, yes most models support decoding with a LM (shallow fusion). For the second one, it was disabled recently because of some changes and need to be updated.

**Comment by lnalborczyk** [comment]
Sorry for not being clear enough. Here is some context.

The context is the same as detailed at the top of this issue. I have trained on network (and a LM) on a large French ASR corpus.

Now I want to use both 1) the decoder from previous stage to help decoding a smaller corpus and 2) use the LM trained at previous stage to help with the decoding of the smaller corpus.

Thanks to you, I now know how to do 1), but I am wondering how to do 2)...

**Comment by b-flo** [comment]
>  use the LM trained at previous stage to help with the decoding of the smaller corpus.

It seems you want to decode a small corpus with a LM trained on another big corpus? If so, you only need to add `--rnnlm [lm_path]` (or `--word-rnnlm ...`) argument to ` asr_recog.py` in your run script.
If you want to do transfer learning between LM, sorry I'm not sure we are supporting the operation. But it would be easy to do as finetuning methods are model independent (see: https://github.com/espnet/espnet/blob/master/espnet/asr/pytorch_backend/asr_init.py), only requirement is to have equivalent architecture.

**Comment by lnalborczyk** [comment]
> It seems you want to decode a small corpus with a LM trained on another big corpus? If so, you only need to add --rnnlm [lm_path] (or --word-rnnlm ...) argument to asr_recog.py in your run script.

That's exactly what I am trying to do, thanks!

**Comment by lnalborczyk** [comment]
Sorry to revive this thread, but is there a way in ESPnet to transfer the decoder (as discussed abode) and to freeze the weights of the decoder for a given period of time (e.g., for the first ten epochs), before to resume to "normal" training?

**Comment by Ryuk17** [comment]
> enc-init: 'exp/tr_it_pytorch_train_rnn/results/model.acc.best'
> enc-init-mods: 'enc.0'
> dec-init: 'exp/tr_it_pytorch_train_rnn/results/model.acc.best'
> dec-init-mods: 'dec.1'

Thank you very much for your detailed answer! 
What's the meaning of the '0' in 'enc.0' and '1' in dec.1?

**Comment by b-flo** [comment]
That's the layer number (from 0 to N-1)!
